### PR TITLE
Add procmatch library

### PR DESCRIPTION
**Pull request by sfluor**
### What does this PR do?

Adds a process matching library that allows to extract integrations from command line processes

### Motivation

Allow integration discovery on the agent-side on the long term

### Additional Notes

Anything else we should know when reviewing?

### CODE REVIEW

File: pkg/procmatch/contains_matcher.go
Comment (sunhay) [inline pkg/procmatch/contains_matcher.go]: We found that there was a high rate of false positives with the `containsMatcher`, so lets remove this and only leave the graph-based matcher

File: pkg/procmatch/readme.md
Comment (sunhay) [inline pkg/procmatch/readme.md]: This can be simply `NewMatcher()`, once the contains-based search is removed. No need to leak implementation details.

File: pkg/procmatch/procmatch_test.go
Comment (sunhay) [inline pkg/procmatch/procmatch_test.go]: `TestFindService` -> `TestFindIntegration`

File: pkg/procmatch/cmd/procmatch/procmatch.go
Comment (sunhay) [inline pkg/procmatch/cmd/procmatch/procmatch.go]: Incorrect import. Do we still need this repo if we're just embedding the library within this agent?

File: pkg/procmatch/cmd/procmatch/procmatch.go
Comment (conorbranagan) [inline pkg/procmatch/cmd/procmatch/procmatch.go]: Do we actually need a new command bundled with the agent? If we want it for testing then I'd rather make it some sub-command of the `agent` binary itself and have it support a bit more (show me all detected integrations on my current host, ...)

File: pkg/procmatch/graph_matcher_test.go
Comment (conorbranagan) [inline pkg/procmatch/graph_matcher_test.go]: nit: I don't think this comment is needed

File: pkg/procmatch/procmatch.go
Comment (conorbranagan) [inline pkg/procmatch/procmatch.go]: Adding a panic seems very risky, especially for a customer-side application. That will mean if there's some odd bug introduced we could potentially block the whole Agent from starting when it's just the process matching that's broken.

I'd prefer this to log a big error and then have some handling of an uninitialized `Matcher`

File: pkg/procmatch/procmatch_test.go
Comment (conorbranagan) [inline pkg/procmatch/procmatch_test.go]: If we're only splitting these up for the benchmarks I think I'd lean towards keeping a single list and just grabbing pieces of the slice. This will make it clearer for the next person adding integrations with tests. I'd also recommend sorting all the integrations.

File: pkg/procmatch/readme.md
Comment (conorbranagan) [inline pkg/procmatch/readme.md]: I think in this documentation I'd like to see step-by-step process on:

1. Adding a new integration + signatures
2. Adding tests for an integration + signature set

I think what you have already covers some of this but it'd be worth stating these use-cases clearly. I'm also not convinced that we would have a case where someone just installs the procmatch package/command separately.

**Review by conorbranagan — CHANGES_REQUESTED** [review]
Cool, this is a good first pass. I added some thoughts and comments

File: pkg/procmatch/procmatch.go
Comment (sfluor) [reply pkg/procmatch/procmatch.go]: We could remove the `init` function and the `Match` and provide a `NewDefault` that would use the default catalog:

```golang
func NewDefault() (Matcher, error) { return NewMatcher(DefaultCatalog) }
```

This way the error handling would be in the "library client" side

**Review by conorbranagan — DISMISSED** [review]
Just added a couple small nits. From my perspective this looks great for merge!

I'd like to add some sub-commands to the agent binary to match stuff but that can come in a follow-up PR since it's different enough from just adding the library.

**Comment by sfluor** [comment]
I just made a change to use the integrations `display_name` instead of `name`.

**Comment by conorbranagan** [comment]
@sfluor Can we merge this after the freeze on 8/31?

**Comment by sfluor** [comment]
LGTM 👍

**Review by truthbk — COMMENTED** [review]
Small nit, but this looks quite alright to me! :)

**Comment by sfluor** [comment]
Closing then reopening the PR to retrigger appveyor check

**Review by conorbranagan — DISMISSED** [review]
LGTM

**Comment by olivielpeau** [comment]
@sfluor: haven't reviewed the code, but could you add `@DataDog/burrito` as the owner of this new package in the [CODEOWNERS file](https://github.com/DataDog/datadog-agent/blob/master/.github/CODEOWNERS)

I'll find someone in Agent core to review this one final time

File: pkg/procmatch/gen/generate_catalog.go
Comment (olivielpeau) [inline pkg/procmatch/gen/generate_catalog.go]: nit: played a bit with the `go generate` command and I noticed it generated a file with weird indentation, it appears the template here was changed since you generated the `default_catalog.go`, could you fix the template? (for example: unnecessary tab before `"{{$manifest.Name}}"`

File: pkg/procmatch/readme.md
Comment (olivielpeau) [inline pkg/procmatch/readme.md]: nit: like the instructions on `Running the tests` below, the generation command could simply be `INTEGRATIONS_CORE_DIR=<path_to_integration_repo> go generate ./...` (or `INTEGRATIONS_CORE_DIR=<path_to_integration_repo> go generate ./pkg/procmatch/` if we assume we're at the repo root

**Review by olivielpeau — COMMENTED** [review]
Just a couple of nits, but looks good overall!

Ping me when the nits are addressed and I'll ✅

**Review by olivielpeau — APPROVED** [review]
🚀

**Pull request merged**
